### PR TITLE
[BEAM-3865] Fix watermark hold handling bug.

### PR DESCRIPTION
**Pull request by robertwb**
Old watermark must be cleared in case it was wrong. Also strengthen tests to catch this. 

------------------------

Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] Make sure there is a [JIRA issue](https://issues.apache.org/jira/projects/BEAM/issues/) filed for the change (usually before you start working on it).  Trivial changes like typos do not require a JIRA issue.  Your pull request should address just this issue, without pulling in other changes.
 - [ ] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue.
 - [ ] Write a pull request description that is detailed enough to understand:
   - [ ] What the pull request does
   - [ ] Why it does it
   - [ ] How it does it
   - [ ] Why this approach
 - [ ] Each commit in the pull request should have a meaningful subject line and body.
 - [ ] Run `mvn clean verify` to make sure basic checks pass. A more thorough check will be performed on your pull request automatically.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

### CODE REVIEW

**Comment by robertwb** [comment]
R: @charlesccychen

File: sdks/python/apache_beam/transforms/trigger_test.py [hunk 1]
```diff
@@ -81,16 +84,8 @@ def bundle_data(data, size):
 
     if not groupings:
       groupings = [1]
+    grouping = list(groupings) + [-group_by for group_by in groupings]
```
Comment (charlesccychen) [inline sdks/python/apache_beam/transforms/trigger_test.py]: What is this?  This is currently a no-op.

Comment (robertwb) [reply sdks/python/apache_beam/transforms/trigger_test.py]: It expands grouping by also adding negative values, which also tests adding elements in reverse order.

**Comment by robertwb** [comment]
Jenkins: retest this please

File: sdks/python/apache_beam/transforms/trigger_test.py (hunk 1 above)
Comment (charlesccychen) [reply sdks/python/apache_beam/transforms/trigger_test.py]: What I mean is that the variable `grouping` (without "s" at the end) is never used in the rest of the function.  Is this a typo?

Comment (charlesccychen) [reply sdks/python/apache_beam/transforms/trigger_test.py]: What I mean is that the variable `grouping` (without "s" at the end) is never used in the rest of the function.  Is this a typo?

Comment (robertwb) [reply sdks/python/apache_beam/transforms/trigger_test.py]: Oh, yes, nice eye!

And fixing this makes some of the non-deterministic ones fail... Enabling only explicitly on some tests.

Comment (charlesccychen) [inline sdks/python/apache_beam/transforms/trigger_test.py]: > **robertwb** wrote:
> Oh, yes, nice eye!
> 
> And fixing this makes some of the non-deterministic ones fail... Enabling only explicitly on some tests.


Thanks!  Could you also add a comment at the beginning of this method on the behavior of `groupings`?  It is not obvious without a close reading.

File: sdks/python/apache_beam/transforms/trigger_test.py [hunk 2]
```diff
@@ -65,10 +65,21 @@ class TriggerTest(unittest.TestCase):
   def run_trigger_simple(self, window_fn, trigger_fn, accumulation_mode,
                          timestamped_data, expected_panes, *groupings,
                          **kwargs):
+    # Groupings is a list of integers indicating the (uniform) size of bundles
+    # to try. For example, if timestamped_data is has elements [a, b, c, d, e]
```
Comment (charlesccychen) [inline sdks/python/apache_beam/transforms/trigger_test.py]: "is has"

**Pull request merged**